### ISSUE TITLE
Intellisense is very slow.

### ISSUE THREAD

**Issue by davidatbinarystream** [microsoft/AL#7190]
- Issue Type: `Bug`
- Extension Name: `al`
- Extension Version: `10.0.687650`
- OS Version: `Windows_NT x64 10.0.19043`
- VS Code version: `1.71.2`

:warning: We have written the needed data into your clipboard. Please paste! :warning:

When typing Ctrl + Space it takes at least 30 seconds for any intellisense suggestions to pop up, if they open at all.

**Comment by erikrijn** [comment]
Confirmed.
And it's not only the intellisense. VS Code slows down to a crawl with large workspaces open.
When you roll back to the last v9 build, everything is fine again.

**Comment by BernhardKloibmueller** [comment]
Maybe it has something to do with the rule AA0241
Deactivating this role makes it a little bit faster again

The rule now checks the variable types and it looks to me that this is new.

**Comment by fgilsc** [comment]
Confirmed, after upgrading AL Code to v10.0.687650 VS Code became too slow. I had to downgrade to the previous version and it works fine.

**Comment by davidatbinarystream** [comment]
Thanks for the update Erik, let me know if you need any information that will make it easer to debug.

**Comment by davidatbinarystream** [comment]
Thanks Bernhard, I will try that and let you know if that helps.

**Comment by dnpb** [comment]
VS Code overall is nearly unbearable for a few of us, so slow and eventually stops code-analysing properly (or at any reasonable speed) and has to be restarted _even more than normal_ just to be usable for some minutes at a time :-(

**Comment by LoanVision** [comment]
Unfortunately @BernhardKloibmueller suggestion doesn't help :(

**Comment by BernhardKloibmueller** [comment]
@rvanbekkum, @LoanVision 
I am sorry it seems that I was experiencing something different. Due to the other problems with v10.* we downgraded our compilers after I wrote the message so I was not able to verify my own suggestion in daily work.
I was certain it was improving the work after deactivating, I was wrong, sorry for the misleading info

**Comment by BazookaMusic** [comment]
Does disabling code analysis have any effect on the speed of intellisense and the projects in general?

**Comment by NKarolak** [comment]
Has anyone tried **dis**abling the code actions?
https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-code-actions#to-enable-al-code-actions

**Comment by FSchneiderYav** [comment]
Disabling code analysis and disabling code actions both do not result in speedup.

**Comment by sigurpallviggo** [comment]
Can confirm as well. I am not able to use version 10.0.687650 due to it being incredibly slow.

**Comment by BazookaMusic** [comment]
Could you share a few more details?
Some questions:

1.  Is VSCode taking up all the available memory in the last VSIX? Can you share the memory consumption of VSIX when the slowdown is observed?
2. Is the CPU stuck at 100%?
3. Does this happen during project startup or is it consistent after the project has fully loaded?
4. Are there particular scenarios or setups that trigger the slowdown?

**Comment by LoanVision** [comment]
1 - 3-4Gb, not all memory
2 - CPU spikes to 30% after you start typing in something. If app is idle then it goes down to 0%
3 - This is consistent
(4-5) - Our project is 1200 files, with 20-50 files project it's not so noticeable

Machine specs: 24 core CPU/ 64Gb RAM

**Comment by rvanbekkum** [comment]
To add, my experience: It also happens for 'smaller' projects with only around 100 .al files/objects.
CPU goes to 100% after some time, memory usage is moderately high, but not all memory is being used.

**Comment by BazookaMusic** [comment]
Two more questions just to ensure we have the correct information:
1) Could you share your project settings? Are you using incremental build?
2) Could you ensure that when changing settings, you reload the projects? To try with codeanalysis or other changes disabled, the project needs to be reloaded. This has to be done by either restarting VSCode or using reload window from the VS commands.

**Comment by davidatbinarystream** [comment]
HI Sotiris;

I am developing extensions for Business Central using the AL language, so I don't think the first question relates because there is no incremental build and no project settings other than the launch.json and app.json files, which I can send you if you think they would be useful.

I tried un-installing and re-installing VS code.  I also started with a clean checkout of the project I am working on and regenerated the symbols files for AL (AL: Download Symbols).  I did not delete my system or personal settings.  After re-installing (1.71.2) I observed the same performance issues.

Next I un-installed VS code and then installed the previous release (1.70.2) see the screenshot below for additional details.  I experienced some initial problems with this version, but I did not record the details.  I was able to resolve any issues I was having with the 1.70.2 branch by deleting all of my User and Workspace settings (%APPDATA%\Code\User\settings.json).

***@***.***

If there is any other information you need or anything I can do to help please let me know.

**Comment by davidatbinarystream** [comment]
Answers inline below...

**Comment by BernhardKloibmueller** [comment]
@BazookaMusic 

Yes there is a difference when enabling CodeAnalysis, but even deactivated the new version is far slower and opening for instance the Microsoft Base Application Source files which you could also try is taking significant longer.
And navigating in the source files takes longer and the CPU usage is higher, I would say roughly about a half more.
With navigation I speak of Go to Defintion, Find All References and so on. 

When using code analysis I used all 4 analysers in this test activated simultaniously. 

The sub process Microsoft.Dynamics.Nav.EditorServices.Host from VS Code was the one using the CPU and memory.
I hope this information helps you identifying the problem

**Comment by salgiza** [comment]
FWIW: this also happens in macOS.

Our whole team (both Windows and macOS developers) has been suffering from this bug since we upgraded the v10 of the AL extension. We've also had to downgrade to v9 for now (thankfully we were not yet targeting BC21), because VS Code was nearly unusable.

**Comment by RafaelKoch** [comment]
May or may not be related: 

Our product development is seeing no issues, all on latest AL and working against BC21 since mid-October now. Preview Versions and later GA versions used.
At the same time we have internal issues being opened by project teams experiencing just what is described here. 

At the same time I have observed dramatically longer pipeline runs, as soon as we load dependency packages for compilation, that are runtime packages. Opposed to that loading the same as app packages leads to 10% of the compile time. So: load libs as runtimes = looong compile, load libs plain app packages = normal compile time.

Looking at pipeline logs from before AL 10, it appears that using runtimes did not increase the compile time anywhere that dramatically.

So, maybe: issue with symbols loaded from runtimes or runtimes used directly in local packages folder? (In troubleshooting process on our side, don't know yet)

**Comment by BazookaMusic** [comment]
I've been investigating this issue by using the Base Application as a reference and unfortunately it's not trivial to reproduce on my machine. The findings I have until now are:
1) Analyzers make a huge difference in performance, even in small projects. I would definitely suggest not running with background code analysis on large projects because the current implementation saturates resources and can make everything slower. This can be tweaked by setting "al.backgroundCodeAnalysis": false in the "settings.json" file of your project (.vscode folder).
2) We identified a bug where just opening a file in VSCode would trigger a re-compilation of the entire project. This may explain why in large workspaces everything becomes slow because just navigating to different files will force a background compilation. Combined with analyzers, this can tank performance.
3) Using the base application with our default project settings, does not reproduce the problem on my end.

It would be very helpful if you could provide us with a few helpful artifacts so we can see what's happening on your specific machines. Here's a few things that would be very useful:

**Can be shared in github:**
1) Settings.json: Your settings.json file from your project. If it is not there, the global one you are using from (%APPDATA%\Code\User\settings.json). Try enabling '"al.incrementalBuild":  true' and see if that makes a difference.

2) In settings.json, set "al.editorServicesLogLevel" to "Verbose" and run your scenarios. You can find the produced log in:
C:\Users<username>.vscode\extensions\ms-dynamics-smb.al-\bin\win32
The log grows fast so it would be good to reproduce a short scenario.

3) The target runtime version of your projects from app.json

**The items below must be sent privately:**
In order of usefulness:

3) Ideally a project that would reproduce the issue

4) A cpu trace of the process while trying your scenarios. Here's how you can get that:
Download perfview.exe from the official link: [https://github.com/Microsoft/perfview/releases](https://github.com/Microsoft/perfview/releases)
Open perfview.exe and click Collect->Collect at the top bar. In the popup window, set circular MB to be something like 1000MB (the size of the trace) and click collect. The tool will start collecting stack traces. 
![image](https://user-images.githubusercontent.com/36576941/197167861-19bc19ed-55c5-4b51-b885-5bd6b32bbc60.png)

Then you can use VSCode to reproduce the scenarios. 

5) A memory dump of the editor services process. To do that, open task manager and find the vscode process.  Expand it to see the Microsoft.Dynamics.Nav.EditorServices.Host process.
![image](https://user-images.githubusercontent.com/36576941/197168315-06bab03d-179f-4a62-9fd5-5d3f64316be0.png)

To send the private files, I would suggest uploading them to a onedrive account (due to their size) and sending me a private message on Yammer (my name is on my profile).

**Comment by salgiza** [comment]
Ok, this is a bit silly because I posted my comment not even 24 hours ago but... in our case, the problem seems to have been fixed in yesterday's update (v10.0.687650).

**Comment by NKarolak** [comment]
@salgiza Which new version are you referring to ...?
![image](https://user-images.githubusercontent.com/34504100/197190690-1aa7baa3-b83c-4808-8e00-bbdc3173d0e2.png)

**Comment by salgiza** [comment]
@NKarolak Ok, this is really weird but...

![CleanShot 2022-10-21 at 14 37 35](https://user-images.githubusercontent.com/6630306/197197550-68e05d25-0a10-4e10-9f0d-6affd9ce2dba.png)

I've been playing with it, and it seems like VS Code shows the last date _you_ updated/installed the extension, instead of the last date the extension was updated (as it does the same date in the VS Marketplace).

I trusted the date when I saw it this morning, and I thought that it was a new version...

But this makes the bug even weirder. Why did the extension behave correctly today after two weeks of awful performance? 🤔 (Same computer, same project, for me and a coworker I told to update today after I did) Maybe we opened fewer files? (Thinking of the bug that @BazookaMusic mentioned).

Oh well, we'll see what happens on Monday...

**Comment by mjmatthiesen** [comment]
I can confirm that performance has tanked significantly with v10. Going back to 9 works great, but then I don't get the new warnings. Turning off background code analysis works, but then I have to wait on a build to get most of my diagnostics.

**Comment by LoanVision** [comment]
Guys,
with all respect to what you do - this needs to be fixed ASAP. Can we get a status when this will be fixed or what can be temporary fix apart from using old extension? Surely when you open your base app you will see huge lag/slowness.  Is there any ETA?

Thank you.

**Comment by SotirisDragonas** [comment]
@LoanVision Unfortunately, we don't see this slowness in BaseApp and we have not had any complaints from engineers working on about a performance degradation, so we have no leads to investigate. As mentioned earlier in this thread, there's even large third-party projects which are not experiencing this performance slowdown.

**I have compared the performance of previous version and the current version on the BaseApp (multiple versions of baseapp from 19 to 21) and other projects and was not able to reproduce the slowness**. 
I have also compared the performance of analyzers from previous versions with those of the current version and there's no obvious performance difference.

In my previous message, I provided a few ways that you can provide help by giving us profiler traces or logs of what is happening on your machines. It is possible that the specific structure of the projects is causing compilation to take longer.

I still haven't received any of these debugging resources from anyone. If nobody provides these, fixing this is a shot in the dark, thus we will close this issue and promise to improve the general performance. 

In the meantime, I am waiting for someone to reach out to provide these resources and investigating on general performance improvements, which are not guaranteed to solve the issue because again, we can't reproduce it.

**Comment by BazookaMusic** [comment]
@LoanVision Unfortunately, we don't see this slowness in BaseApp and we have not had any complaints from engineers working on about a performance degradation, so we have no leads to investigate. As mentioned earlier in this thread, there's even large third-party projects which are not experiencing this performance slowdown.

**I have compared the performance of previous version and the current version on the BaseApp (multiple versions of baseapp from 19 to 21) and other projects and was not able to reproduce the slowness**. 
I have also compared the performance of analyzers from previous versions with those of the current version and there's no obvious performance difference.

In my previous message, I provided a few ways that you can provide help by giving us profiler traces or logs of what is happening on your machines. It is possible that the specific structure of the projects is causing compilation to take longer.

I still haven't received any of these debugging resources from anyone. If nobody provides these, fixing this is a shot in the dark, thus we will close this issue and promise to improve the general performance. 

In the meantime, I am waiting for someone to reach out to provide these resources and investigating on general performance improvements, which are not guaranteed to solve the issue because again, we can't reproduce it.

**Comment by BernhardKloibmueller** [comment]
@BazookaMusic 

Sorry I did not see your request. 

Then I will provide as much as possible. Due to proper comparison I used the Base Application OnPrem 20.4
Secondly I used a small PTE.

The first thin I noticed is that the size of the log of an the compiler V10 is doubled the size for the same action.

The following Actions were made by me:

Base Application: Opening the folder and openening the file AADApplication.Table.al
My PTE: Openend a Codeunit and adding a procedure saving it and deleting the procedure and saving it.

Unfortunately openening the Base Application will create a 154MB file in Version 9.5 and a 306MB file in Version 10. So no upload is possible.
![grafik](https://user-images.githubusercontent.com/46522333/199521010-9330f4d0-3c92-4feb-90b0-ad50588006de.png)


So I can only send you my small PTE:
Version 9.5:
[_ForGitHub_9.5_EditorServices_ContiniaInteg_Open_NewProcedure.log](https://github.com/microsoft/AL/files/9921383/_ForGitHub_9.5_EditorServices_ContiniaInteg_Open_NewProcedure.log)
Version10:
[_ForGitHub_10_EditorServices_Continia_Open_NewProcedure.log](https://github.com/microsoft/AL/files/9921384/_ForGitHub_10_EditorServices_Continia_Open_NewProcedure.log)

Settings.json:
[_ForGitHub_settings.txt](https://github.com/microsoft/AL/files/9921408/_ForGitHub_settings.txt)
App.json
[_ForGitHub_app.txt](https://github.com/microsoft/AL/files/9921424/_ForGitHub_app.txt)

I have everything set up and additional extensions activated just like in my projects. Maybe some of them is the issue I will get again in touch with you.

I will try to reproduce this with tracefiles and so on but even reproducing this in a clean environment takes time for us so please also consider this that we maybe from your point of view lazy and not giving you enough information to investigate.
Also I will repro everything again when I am able to remove unnecessary extensions.

**Comment by BernhardKloibmueller** [comment]
3 more files to check: With all combinations of Analyzers and deactivated all other extensions in VS Code:

[_ForGitHub_10_EditorServices_DeactivatedAllOtherExtensions_WithAnalyzersActive.log](https://github.com/microsoft/AL/files/9921766/_ForGitHub_10_EditorServices_DeactivatedAllOtherExtensions_WithAnalyzersActive.log)
[_ForGitHub_10_EditorServices_DeactivatedAllOtherExtensions_WithoutAnalyzers.log](https://github.com/microsoft/AL/files/9921767/_ForGitHub_10_EditorServices_DeactivatedAllOtherExtensions_WithoutAnalyzers.log)
[_ForGitHub_10_EditorServices_DeactivatedAllOtherExtensions_WithoutBackgroundAnalyzers.log](https://github.com/microsoft/AL/files/9921768/_ForGitHub_10_EditorServices_DeactivatedAllOtherExtensions_WithoutBackgroundAnalyzers.log)

It looks like the analyzers are running wild in V10. This is only due to the fact that from my point of view the size of the log files is indicating that VS Code or the editor service is working more when logging more.

**Comment by BazookaMusic** [comment]
@BernhardKloibmueller Thank you very much for the logs. It's very interesting that they're larger and a good first step for seeing what the compiler is doing. I understand that getting the information is time consuming, I was just trying to explain why the problem is hard to diagnose.

With a little bit of python magic, I analyzed all the logs and I can see that in the newer version, there's an order of magnitude more code actions being executed. For example, in the two first logs you provided, version 10 executes **7685 code actions more than version 9.5.** (which is ~7 times the amount of actions run by 9.5)

Looking at your settings, I noticed this:
    "editor.codeActionsOnSave": {
        "source.fixAll": true
    }
 New code actions were added in version 10 and it might be that this setting is causing them to be run in a cascade. **I would try to see if disabling code actions helps in this case.**
 
 It could also be a bug where we are running code actions when we shouldn't, which is hard to judge from these logs but at least a good lead.

**Comment by fgilsc** [comment]
Any update about this?

**Comment by dnpb** [comment]
Yes, right above... Please, don't let's all post such noisy comments, which just lead to more emails to delete for lots of people.

**Comment by BernhardKloibmueller** [comment]
@BazookaMusic 
I have sent you on Yammer a private message with the large files and my project folder for you to investigate

**Comment by fgilsc** [comment]
> 9.X release :)

Yeah... but it's a bit tedious, uninstall release 9 and install release 10 when I want to compile.
We need a solution.

**Comment by mjmatthiesen** [comment]
> > 9.X release :)
> 
> Yeah... but it's a bit tedious, uninstall release 9 and install release 10 when I want to compile. We need a solution.

You do not need to install 10 to compile. Just use the compile from 9.x. It is not an issue at all. I use 9.x to deploy to sandboxes, my azure+docker vms, and so forth. Any app that actually goes to a CX goes through a pipeline that doesn't need to worry about the awful performance.

**Comment by fgilsc** [comment]
> You do not need to install 10 to compile. Just use the compile from 9.x. It is not an issue at all. I use 9.x to deploy to sandboxes, my azure+docker vms, and so forth. Any app that actually goes to a CX goes through a pipeline that doesn't need to worry about the awful performance.

If you app runtime is 10, you won't able to compile with 9.x...

**Comment by davidatbinarystream** [comment]
HI Sotiris;

I have some information that might be helpful...

I downgraded my VS code to the previous (1.71) release, but I was having some problems running it, so I deleted all my configuration files and recreated them, which allowed me to use the 1.71 version without any issues.

Recently, I allowed VS Code to update itself to 1.72.2, and I am no longer experiencing the performance issues.    My guess is that there is some conflict between the new version VS Code (1.72.2) and the various configuration files created by 1.71.  Unfortunately, I did not save my previous configuration files, but I would bet that if you could find and installation that was exhibiting the performance problems and cross reference the configuration files with a healthy system you would find your smoking gun.

Best regards, Dave

**Comment by BazookaMusic** [comment]
@davidatbinarystream Thank you for sharing this. By configuration settings, I assume you mean the VSCode settings in %APPDATA%\Code\User\settings.json.

**It would be quite helpful if whoever is affected by this could try to take a backup of their VSCode settings and try to regenerate them and see if that fixes the issue.

If you could also share the configuration files here, it would be very helpful.**

On the files shared by Bernhard:

The defining difference between the versions seemed to be that we were compiling more often due to code actions in version 10. Setting "al.enableCodeActions":false should prevent that. It's not a good idea to enable code analyzers unrelated to your fix if you want to use code actions for now. We only have actions for CodeCop, but we **are running all analyzers** on every click after every change to discover possible code actions. This is a bug that existed for v9 as well however and also we have claims that disabling them did not help.

If the issue is on VSCode as Dave pointed out, that would explain why I cannot repro it. My VSCode is on version **1.73.**

**Comment by BernhardKloibmueller** [comment]
@BazookaMusic 
My VS Code Version was:
Version: 1.72.2 (system setup)
Commit: d045a5eda657f4d7b676dedbfa7aab8207f8a075
Date: 2022-10-12T22:15:18.074Z
Electron: 19.0.17
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Windows_NT x64 10.0.19044
Sandboxed: No

**Comment by BernhardKloibmueller** [comment]
@BazookaMusic 
[_ForGitHub_LargerProject_EditorServices.zip](https://github.com/microsoft/AL/files/9952822/_ForGitHub_LargerProject_EditorServices.zip)

I have attached a larger project where the issue is significant

I could now reproduce it the following:
I opened the project
Cut 1 line
Pasted the 1 line on the exact same position
Saved the file
Saving takes a few seconds

Do you see anything in the logs or do you again need a comparison or dump files?
I ask because it would take again some hours to reproduce everything

**Comment by BazookaMusic** [comment]
@BernhardKloibmueller Is it maybe AL Format that is causing this slowdown? This can be caused by having format on save enabled (in the settings).  I have seen that AL format can be problematic on the Base Application, where it can take multiple seconds . Saving a file is not an action that involves the extension, unless there's a formatting step before saving.

This is also something that sounds unrelated to the original issue of intellisense being slow and general slowness in the UI.

**Comment by BernhardKloibmueller** [comment]
@BazookaMusic 
Yes I have Format On Save in AL activated.

I have now tried the following combinations:
1. Fast
 - FormatOnSave Deactivated
 - CodeActoinsOnsave Deactivated
 - CodeAnalysis Deactivated

2. Fast
 - FormatOnSave Activated
 - CodeActoinsOnsave Deactivated
 - CodeAnalysis Deactivated
 
3. Fast
 - FormatOnSave Activated
 - CodeActoinsOnsave Activated
 - CodeAnalysis Deactivated

4. Slow
 - FormatOnSave Deactivated
 - CodeActoinsOnsave Deactivated
 - **CodeAnalysis Activated**

5. Slow
 - FormatOnSave Activated
 - CodeActoinsOnsave Deactivated
 - **CodeAnalysis Activated**

6. Slowest
 - FormatOnSave Activated
 - CodeActoinsOnsave Activated
 - **CodeAnalysis Activated**

So it seems indeed that it is unnecessary if FormatOnSave is activated or CodeActions are activated, If the CodeAnalysis is activated then it is slow. But why?

My Test was: same action as the last log I sent you. I removed a line, readded the line, went straight to the next line and started typing the next word I wanted to be suggested in Intellisense. It took with CodeAnalysis some seconds to respond to the typing.
Do you need a new log?

I have taken a screen video to further illustrate my problem with the current situation.
https://user-images.githubusercontent.com/46522333/201025430-6c2353b2-3866-40b9-bd24-a09b177457a4.mp4

https://user-images.githubusercontent.com/46522333/201025439-7d07fabe-5299-41fc-8e74-99cc88ec00fb.mp4

**Comment by BazookaMusic** [comment]
Unfortunately, if it's code analysis that is causing the issue, there isn't a trivial solution to this problem. The good news is that we are working on improving the analyzers in this release to run incrementally, thus removing a huge amount of the performance issues you are seeing. This is however ongoing work and there isn't a simple fix we can provide for this because their implementation is flawed.

The slowdown you are seeing is because background code analysis is enabled. The extension is trying to run analyzers on the entire project during startup. A way to avoid this but still keep analyzers for your project is to set:
**"al.backgroundCodeAnalysis": false**
This will mean that the analyzers will fully run only when you build.

**Another thing to consider, to actually disable code actions you need to do:
 **"al.enableCodeActions": false****

To discover the code actions, we have a sub-optimal implementation where we were running all analyzers on the entire project. This can mean that we will be running all the analyzers on the entire project every time VSCode tries to discover code actions, which is almost on every click. For large projects, it's impossible to work with this.  For this reason, I have added this to the ongoing effort to improve the analyzers.

Furthermore, with the next release of the VSIX which will be soon released, we have removed a large amount of useless analyzer runs (the bug I mentioned), so you should see some improvement short term.

To summarize, we are working on improving the analyzers in this release and you should see significant improvement in the next few releases of the extension. The issue is a general performance problem with the analyzers.

For now and until we refactor their implementation, the mitigations are by order of impact:
1)  **Set **"al.enableCodeActions": false**. This will ensure that code actions will not invoke analysis of the whole project on almost every click.**
2) Set **"al.backgroundCodeAnalysis": false**. This will ensure that analyzers only run on build and don't take away performance during development. 
3) Enable part of the analyzers. Running less of them will make the experience faster.

**Comment by BazookaMusic** [comment]
As mentioned, the improvements will come in the next releases of the VSIX because we will need to test them. With this, I will close this issue.

**Comment by fartwhif** [comment]
Having this problem, had to use the mitigation.

AL Language Version:
`Version: 12.1.883011`
VSCode Version:
```Version: 1.80.1 (system setup)
Commit: 74f6148eb9ea00507ec113ec51c489d6ffb4b771
Date: 2023-07-12T17:22:07.651Z
Electron: 22.3.14
ElectronBuildId: 21893604
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Windows_NT x64 10.0.19045```

**Comment by BazookaMusic** [comment]
Try following this guide to figure out what's slow on your project.
https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-code-analysis-performance-configuration

Compile with the flag that gives you the analyzer statistics and disable any rules which are too slow on your machine. Also which part of the mitigation helped the most? Is it code actions or disabling the entire code analysis?
